doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,279 @@
+# AWS Connection
+
+## The Problem
+
+Github issue:  [Issue 7256](https://github.com/MaterializeInc/materialize/issues/7256)
+
+As an operational data warehouse, Materialize sits upstream of a traditional data warehouse
+for our customers. The data brought in to Materialize will serve not only operational workloads,
+but ultimately analytical ones. This means our customers can benefit from a cheap and
+user-friendly method to share their data or send it on to the tools purpose-built for
+their analytical workloads while reducing the need to duplicate transformation logic
+across operational and analytical data warehouses. A full copy of data from Materialize
+into S3 is a simple and intuitive way to enable this for our customers.
+
+## Goals
+* Allow batch exports of data in Materialize to S3 in a format that can be easily ingested into analytical data warehouses.
+* Provide basic internal and external observability into the frequency, reliability, and size of the batch exports.
+* Tightly constrain scope so that the initial feature can reach public preview in early Q1 next year.
+
+## Non-goals
+* Supporting continuous or scheduled exports.
+* Integrating directly with batch data warehouses, without writing to S3.
+
+## Solution Proposal
+
+### SQL
+In materialize we already have some support for `COPY`
+commands. We should expand upon them, keeping them as close to
+the [postgres syntax](https://www.postgresql.org/docs/current/sql-copy.html).
+
+The SQL would look like the following.
+
+```sql
+COPY <mv_name> -- mv or table name
+TO <scalar_expr_for_s3_path> -- the scalar_expr should give a path like 's3://prefix'
+WITH (
+  AWS CONNECTION = aws_conn -- previously created aws connection
+  FORMAT = 'csv', -- file format name
+  MAX FILE SIZE = 16777216 -- max file size in bytes, to limit size of files in s3
+)
+```
+This option is the most similar to the Postgres `COPY` command syntax.
+Though note, there is no `COPY` in the SQL standard and Postgres itself used to support a
+different syntax in the older versions.
+
+For reference, this is what an example SQL looks like in other data warehouses and postgres.
+
+[Redshift](https://docs.aws.amazon.com/redshift/latest/dg/r_UNLOAD.html)
+```sql
+-- Redshift
+UNLOAD ('select * from venue')
+TO 's3://mybucket/unload/'
+IAM_ROLE 'arn:aws:iam::0123456789012:role/MyRedshiftRole'
+CSV DELIMITER AS '|';
+```
+[Snowflake](https://docs.snowflake.com/en/sql-reference/sql/copy-into-location)
+```sql
+-- Snowflake
+COPY INTO 's3://mybucket/unload/'
+  FROM mytable
+  STORAGE_INTEGRATION = myint -- ~ equivalent to a materialize connection
+  FILE_FORMAT = (TYPE = csv FIELD_DELIMITER = '|');
+```
+[Postgres](https://www.postgresql.org/docs/current/sql-copy.html)
+```sql
+COPY table_name
+TO 'path/to/file'
+WITH (FORMAT CSV);
+```
+
+
+#### S3 Path
+For the s3 path, users should be able to provide an expression
+like `'s3://prefix/path/' || mz_now()` to help generate the path. This would
```

Comment:
How is the path determined when multiple files are written (e.g. if the max size is set to a value that is less than the size of the table)?

Is this path a prefix and we have some known format for the per-file suffix?

Reply:
Ah, this s3 path should be an empty directory i.e. just the path prefix. I added a section for s3 file names below, will link here and further clarify.